.github/workflows/iroha2-dev.yml

```diff
@@ -12,84 +13,80 @@ env:
+          tags: iroha2:base
           context: .
-          push: true
-          tags: hyperledger/iroha2:dev
-          build-args: |
-            TARGET_DIR=release
-            PROFILE=--release
-      - name: Build and push docker image (iroha_client_cli)
-        uses: docker/build-push-action@v2
+          file: Dockerfile.base
+
+      - name: Build iroha2:build docker image
+        uses: docker/build-push-action@v3
         with:
+          load: true
+          tags: iroha2:build
           context: .
-          push: true
-          tags: hyperledger/iroha2:client-cli-dev
-          build-args: |
-            TARGET_DIR=release
-            PROFILE=--release
-            BIN=iroha_client_cli
-      - name: Build and push docker image (kagami)
-        uses: docker/build-push-action@v2
+          file: Dockerfile.build
+
+      - name: Build iroha2:dev docker image
+        run: |
+          docker build -f Dockerfile -t hyperledger/iroha2:dev .
+      - name: Login to DockerHub
+        uses: docker/login-action@v2
         with:
-          context: .
-          push: true
-          tags: hyperledger/iroha2:kagami-dev
-          build-args: |
-            TARGET_DIR=release
-            PROFILE=--release
-            BIN=kagami
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      - name: Push iroha2:dev docker image
+        run: |
+          docker push hyperledger/iroha2:dev
 
       - name: Build and push docker image (load-rs:dev)
         run: |
           sleep 10s
           echo "wait to other workflow"
-      - uses: convictional/trigger-workflow-and-wait@v1.6.0
+      - uses: convictional/trigger-workflow-and-wait@v1.6.2
         with:
           owner: soramitsu
           repo: iroha2-longevity-load-rs
           github_token: ${{ secrets.G_ACCESS_TOKEN }}
           workflow_file_name: load-rs-push-from-dev.yaml
           ref: dev
 
-  build-ci-image:
+  upload_coverage:
+    if: github.event.pull_request.merged == true
     runs-on: ubuntu-latest
     container:
       image: 7272721/i2-ci:latest
     steps:
       - uses: actions/checkout@v2
-      - uses: Swatinem/rust-cache@v1
-      - name: Login to DockerHub
-        uses: docker/login-action@v1
+      - name: Download codecov report artifact
+        uses: dawidd6/action-download-artifact@v2.21.1
         with:
-          username: ${{ secrets.DOCKERHUB_USERNAME }}
-          password: ${{ secrets.DOCKERHUB_TOKEN }}
-      - name: Build & push Docker container
-        uses: docker/build-push-action@v2
+          workflow: iroha2-dev-pr.yml
+          name: lcov-${{ github.event.pull_request.head.sha }}.info
+          search_artifacts: true
+      - name: Upload to codecov
+        uses: codecov/codecov-action@v3
         with:
-          file: Dockerfile.ci
-          tags: hyperledger/iroha2-ci:latest
-          push: true
+          file: lcov-${{ github.event.pull_request.head.sha }}.info
```

Comment:
One thing,
if we specify the commit ID in the filename for codecov, does codecov find relation between the different filenames and make proper comparisons?

Reply:
This is actually a good point. I didn't spend time to verify it, anyway, since `github.event.pull_request.head.sha` will be different for base branch lcov report and for merge branch correspondingly, we might unify the filenames for reports to `lcov.info`. Since files for `upload & download artifacts` should be the same, codecov has to join these run IDs reports in one coverage by PR head sha.